### PR TITLE
bpf: avoid race when selecting the RevSNAT port

### PR DESCRIPTION
**Pull request by lmb**
The logic to allocate SNAT mapping contains a race condition. At a high level it does the following:

    if (!revsnat_exists(port)) {
        if (!create_revsnat(port)
            return error;

        ...
    }

Two concurrent executions of the datapath may succeed the revsnat_exists check, which then leads to one of them bailing out since create_revsnat fails.

Instead simply try to create the RevSNAT entry. If that fails we retry with another port.

Updates: #33074

```release-note
Avoid race during RevSNAT mapping creation, resulting in packet drop with "No mapping for NAT masquerade".
```

### CODE REVIEW

File: bpf/lib/nat.h
Comment (julianwiedmann) [inline bpf/lib/nat.h:223]: fwiw, I believe we only reach  `snat_v4_new_mapping()` if the preceding SNAT logic didn't find any SNAT entry for the connection. So at least an `-EEXIST` here would be *very* unexpected.

**Review by julianwiedmann — APPROVED** [review]
Thank you, nice catch!

I slightly word-smithed the release note - but imho this is even a `release-note/bug`, and we should queue up a backport?

**Comment by lmb** [comment]
> but imho this is even a `release-note/bug`, and we should queue up a backport?

I had a look, this is present even in the first version 0e785f9 which appeared in 1.5.0. There were a couple of changes in between, so the backport would have to be manual.

I'll leave it up to you to decide - maybe it makes sense to see whether this helps the user with their symptoms?

File: bpf/lib/nat.h
Comment (lmb) [reply bpf/lib/nat.h:223]: Agreed. Do you want me to do something about this?

Comment (julianwiedmann) [reply bpf/lib/nat.h:223]: I wouldn't touch it as part of this PR, no. Preserving the old behaviour for now makes sense. And in the end we couldn't delete the rollback path - I suspect we would just want to install the SNAT entry without `BPF_NOEXIST`, but still handle unexpected errors.

**Comment by julianwiedmann** [comment]
>I'll leave it up to you to decide - maybe it makes sense to see whether this helps the user with their symptoms?

Ack, sounds good. Let's release-note it as bug, but hold off on backports until we see if it helps.

**Pull request merged**
